crates/bevy_ecs/src/system/query.rs

```diff
@@ -9,246 +9,270 @@ use crate::{
+/// |[`get`]\([`_mut`][`get_mut`])|Returns the query item for the specified entity.|
+/// |[`many`]\([`_mut`][`many_mut`]),<br>[`get_many`]\([`_mut`][`get_many_mut`])|Returns the query items for the specified entities.|
+/// |[`single`]\([`_mut`][`single_mut`]),<br>[`get_single`]\([`_mut`][`get_single_mut`])|Returns the query item while verifying that there aren't others.|
 ///
-/// ```
-/// # use bevy_ecs::prelude::*;
-/// # #[derive(Component)]
-/// # struct ComponentA;
-/// # #[derive(Component)]
-/// # struct ComponentB;
-/// fn immutable_query_system(query: Query<(&ComponentA, &ComponentB)>) {
-///     for (a, b) in &query {
-///         // Here, `a` and `b` are normal references to components, relatively of
-///         // `&ComponentA` and `&ComponentB` types.
-///     }
-/// }
-/// # bevy_ecs::system::assert_is_system(immutable_query_system);
+/// There are two methods for each type of query operation: immutable and mutable (ending with `_mut`).
+/// When using immutable methods, the query items returned are of type [`ROQueryItem`], a read-only version of the query item.
+/// In this circumstance, every mutable reference in the query fetch type parameter is substituted by a shared reference.
 ///
-/// fn mutable_query_system(mut query: Query<(&mut ComponentA, &ComponentB)>) {
-///     for (mut a, b) in &mut query {
-///         // Similar to the above system, but this time `ComponentA` can be accessed mutably.
-///         // Note the usage of `mut` in the tuple and the call to `iter_mut` instead of `iter`.
-///     }
-/// }
-/// # bevy_ecs::system::assert_is_system(mutable_query_system);
-/// ```
+/// # Performance
+///
+/// Creating a `Query` is a low-cost constant operation.
+/// Iterating it, on the other hand, fetches data from the world and generates items, which can have a significant computational cost.
+///
+/// [`Table`] component storage type is much more optimized for query iteration than [`SparseSet`].
+///
+/// Two systems cannot be executed in parallel if both access the same component type where at least one of the accesses is mutable.
+/// This happens unless the executor can verify that no entity could be found in both queries.
+///
+/// Optional components increase the number of entities a query has to match against.
+/// This can hurt iteration performance, especially if the query solely consists of only optional components, since the query would iterate over each entity in the world.
+///
+/// The following table compares the computational complexity of the various methods and operations, where:
 ///
-/// ## Getting the query result for a particular entity
+/// - **n** is the number of entities that match the query,
+/// - **r** is the number of elements in a combination,
+/// - **k** is the number of involved entities in the operation,
+/// - **a** is the number of archetypes in the world,
```

Comment:
I have been trying to figure out whether using a Changed<> filter decreases complexity to the number of changed matching components instead of all (even unchanged) matching components, and this notion of a archetype is new to me. What does this mean exactly, will adding a Changed filter speed up queries or not?

Looking at #5097 seems to indicate that it does not and then this line is maybe worded slightly wrongly. Maybe it should be `a is the number of components in matching archetypes.` as in the mentioned issue?

Seeing as this is a docs change, I think my lacking knowledge is maybe a good thing for improving the quality of the docs.

Reply:
The asymptotic efficiency of Changed filters is (currently) the same as it performs a binary check for each entity. The performance improvements are if the binary check is faster than the operation you would otherwise do.